### PR TITLE
[OCPCLOUD-1647] Implement pre-flight Cluster state validation

### PR DESCRIPTION
**Pull request by damdo**
We need to make sure a number of checks pass before the CPMS controller can perform any operations.
The things to be checked are:

1. **All Nodes in the cluster claiming to be control plane nodes have a valid machine**.
1. **At least 1 of the control plane machines is in the ready state** (if there are no ready Machines then the cluster is likely misconfigured).
1. **The cluster has the correct number of indexes**:
      + Right number of indexes, The cluster state is valid.
      + Too few indexes, The cluster state is valid. We will later scale up without user intervention when we perform reconcileMachineUpdates.
      + Too many indexes, invalid. We set the operator to degraded and ask the user for manual intervention.
1. **No replacement machines** (one that doesn't need update but has an equivalent in the index that needs update) **have an error.**

If any of these conditions is not met, the cpms operator will go degraded and will halt all operations until the issue has been resolved.

### CODE REVIEW

File: pkg/controllers/controlplanemachineset/controller.go [hunk 1]
```diff
@@ -470,3 +529,181 @@ func isOwnedByCurrentCPMS(cpms *machinev1.ControlPlaneMachineSet, machineObjectM
...
+		nodesList := &corev1.NodeList{}
+		if err := r.List(ctx, nodesList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				label: "",
+			}),
+		}); err != nil {
+			return false, fmt.Errorf("failed to get Nodes: %w", err)
+		}
+
+		cpmsNodes = append(cpmsNodes, nodesList.Items...)
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/controller.go:550]: Does this mean we could end up with duplicate nodes? Should we deduplicate this list based on node names?

File: pkg/controllers/controlplanemachineset/controller.go [hunk 2]
```diff
@@ -470,3 +529,181 @@ func isOwnedByCurrentCPMS(cpms *machinev1.ControlPlaneMachineSet, machineObjectM
 
 	return false
 }
+
+func (r *ControlPlaneMachineSetReconciler) checkControlPlaneNodesToMachinesMappings(ctx context.Context, logger logr.Logger,
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/controller.go:533]: Missing description

File: pkg/controllers/controlplanemachineset/controller.go [hunk 3]
```diff
@@ -470,3 +529,181 @@ func isOwnedByCurrentCPMS(cpms *machinev1.ControlPlaneMachineSet, machineObjectM
...
+			Message: fmt.Sprintf("Found %d unmanaged node(s)", len(unmanagedNodeNames)),
+		})
+
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (r *ControlPlaneMachineSetReconciler) checkReadyControlPlaneMachineExists(logger logr.Logger, cpms *machinev1.ControlPlaneMachineSet, sortedIndexedMs [][]machineproviders.MachineInfo) bool {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/controller.go:596]: Missing description

File: pkg/controllers/controlplanemachineset/controller.go [hunk 4]
```diff
@@ -470,3 +529,181 @@ func isOwnedByCurrentCPMS(cpms *machinev1.ControlPlaneMachineSet, machineObjectM
...
+			Message: "No ready control plane machines found",
+		})
+
+		return false
+	}
+
+	return true
+}
+
+func (r *ControlPlaneMachineSetReconciler) checkCorrectNumberOfIndexes(logger logr.Logger, cpms *machinev1.ControlPlaneMachineSet, sortedIndexedMs [][]machineproviders.MachineInfo) bool {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/controller.go:631]: Missing description

File: pkg/controllers/controlplanemachineset/controller.go [hunk 5]
```diff
@@ -470,3 +529,181 @@ func isOwnedByCurrentCPMS(cpms *machinev1.ControlPlaneMachineSet, machineObjectM
...
+			Message: fmt.Sprintf("Observed %d index(es) in excess", excessiveIndexes),
+		})
+
+		return false
+	}
+
+	return true
+}
+
+func (r *ControlPlaneMachineSetReconciler) checkNoErrorForReplacements(logger logr.Logger, cpms *machinev1.ControlPlaneMachineSet, sortedIndexedMs [][]machineproviders.MachineInfo) bool {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/controller.go:669]: Missing description

File: pkg/controllers/controlplanemachineset/controller.go [hunk 6]
```diff
@@ -470,3 +529,181 @@ func isOwnedByCurrentCPMS(cpms *machinev1.ControlPlaneMachineSet, machineObjectM
...
+	var nonReadyReplacementMachineNames []string
+
+	for _, machines := range sortedIndexedMs {
+		machinesUpdated := updatedMachines(machines)
+		machinesOutdated := needReplacementMachines(machines)
+
+		if hasAny(machinesOutdated) && hasAny(machinesUpdated) {
+			for _, m := range machinesUpdated {
+				if m.ErrorMessage != "" {
+					nonReadyReplacementMachineNames = append(nonReadyReplacementMachineNames, m.MachineRef.ObjectMeta.Name)
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/controller.go:679]: Are these nonReady or errored? I think nonReady would be valid no?

File: pkg/controllers/controlplanemachineset/controller.go (hunk 1 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/controller.go]: Yes better. Done.

File: pkg/controllers/controlplanemachineset/controller.go (hunk 6 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/controller.go]: Ah good catch! Updated.

File: pkg/controllers/controlplanemachineset/controller.go [hunk 7]
```diff
@@ -470,3 +529,196 @@ func isOwnedByCurrentCPMS(cpms *machinev1.ControlPlaneMachineSet, machineObjectM
...
+		})
+
+		return false
+	}
+
+	return true
+}
+
+// fetchControlPlaneNodes fetches a map of unique nodes that have the "control-plane" (and/or legacy "master") labels.
+func (r *ControlPlaneMachineSetReconciler) fetchControlPlaneNodes(ctx context.Context) (*map[string]corev1.Node, error) {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/controller.go:707]: Why return a pointer to a map?

Comment (damdo) [reply pkg/controllers/controlplanemachineset/controller.go]: It is superfluous for maps, you are right. Removed

**Comment by JoelSpeed** [comment]
Ok, looks good, please squash and I'll label

**Pull request merged**